### DISCUSSION TITLE
problem with bash script and quotation marks

### DISCUSSION THREAD

**Discussion by Jmuccigr** [ImageMagick/ImageMagick#2879] [Help] [answered]
I'm having a problem with a bash script I'm writing for MacOS and I suspect it's me not understanding something basic about escaping.

So, something that works:

```
cmdfinal="-size 600x800 xc: show:"
magick $cmdfinal
magick -size 600x800 xc: show:
```

I get two blank documents. Now, I'd have thought that quoting the variable wouldn't make a difference, but it does.

```
cmdfinal="-size 600x800 xc: show:"
magick "$cmdfinal"
magick -size 600x800 xc: show:
```
And I get this error: `Error: Invalid argument or not enough arguments`

What I'm really trying to do is a little more complicated, but here's the simplified version:
```
cmdfinal="-size 600x800 xc: -fill white -stroke black -strokewidth 1 -draw \"line 50,35 50,605\" show:"
magick $cmdfinal
magick -size 600x800 xc: -fill white -stroke red -strokewidth 1 -draw "line 50,35 50,605" show:
```
(I've made the stroke color different in each just so I can tell from a glance which output I'm looking at.)

Now the direct command still works, but I get this error from the variable-using one:
> magick: non-conforming drawing primitive definition `line' @ error/draw.c/RenderMVGContent/4496.

Same error if I use single quotes instead.

Quote the variable and we're back to the other error: `Error: Invalid argument or not enough arguments`

I've played around a bit, but I' can't seem to get to a version where the quotation marks are not generating an error.

TIA.

MacOS 10.13.6
Version: ImageMagick 7.0.10-38 Q16 x86_64 2020-11-16 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1)
Delegates (built-in): bzlib fftw fontconfig freetype heic jng jp2 jpeg lcms ltdl lzma png tiff webp wmf x xml zlib

**Comment by fmw42** [comment, marked as answer]
Basically you cannot have a variable that you quote that has quotes in it. But there is a workaround.  See the man pages for eval. This is not a Mac-specific issue. 

So this works

```
cmdfinal="-size 600x800 xc: -fill white -stroke black -strokewidth 1 -draw \"line 50,35 50,605\" show:"
eval magick $cmdfinal
magick -size 600x800 xc: -fill white -stroke red -strokewidth 1 -draw "line 50,35 50,605" show:

```

**Reply by snibgo** [reply]
Cygwin bash on Windows has the behaviour shown in the OP, and fmw42's "eval" workaround works.

**Reply by Jmuccigr** [reply]
Excellent. Thanks. (I had eval in the back of my mind, but hadn't tried it.)

So is this a magick thing, then, or a broader bash issue?

**Comment by fmw42** [comment]
It is bash
